### ISSUE TITLE
[recommendation] protobuf::FatalException error

### ISSUE THREAD

**Issue by christ1ne** [tensorflow/models#4787]
Please go to Stack Overflow for help and support:

http://stackoverflow.com/questions/tagged/tensorflow

Also, please understand that many of the models included in this repository are experimental and research-style code. If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **What is the top-level directory of the model you are using**:  models/official/recommendation
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
pip anaconda  https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-1.9.0-cp34-cp34m-linux_x86_64.whl
- **TensorFlow version (use command below)**: branch r1.9.0
- **Bazel version (if compiling from source)**: 1.9.0
- **CUDA/cuDNN version**: cuda 9.2 / cuDNN7.0.5
- **GPU model and memory**:  Titan Xp, 12192MiB
- **Exact command to reproduce**: python ncf_main.py --model_dir model_20m  --data_dir data --dataset ml-20m

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

code stops running with error messages

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

$ python
Python 2.7.15 |Anaconda custom (64-bit)| (default, May  1 2018, 23:32:55) 

$ python ncf_main.py --model_dir model_20m  --data_dir data --dataset ml-20m
/home/n/anaconda3/envs/p27_cu92/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
2018-07-13 10:52:04.831652: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-07-13 10:52:04.915329: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-07-13 10:52:04.915744: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1392] Found device 0 with properties:
name: TITAN Xp major: 6 minor: 1 memoryClockRate(GHz): 1.582
pciBusID: 0000:01:00.0
totalMemory: 11.91GiB freeMemory: 11.53GiB
2018-07-13 10:52:04.915755: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1471] Adding visible gpu devices: 0
2018-07-13 10:52:05.065370: I tensorflow/core/common_runtime/gpu/gpu_device.cc:952] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-07-13 10:52:05.065407: I tensorflow/core/common_runtime/gpu/gpu_device.cc:958]      0
2018-07-13 10:52:05.065412: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   N
2018-07-13 10:52:05.065610: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1084] Created TensorFlow device (/device:GPU:0 with 11162 MB memory) -> physical GPU (device: 0, name: TITAN Xp, pci bus id: 0000:01:00.0, compute capability: 6.1)
I0713 10:52:05.157206 139797621028608 tf_logging.py:115] Data preprocessing...
I0713 10:53:16.772550 139797621028608 tf_logging.py:115] Creating Estimator from Keras model...
I0713 10:53:17.262474 139797621028608 tf_logging.py:115] Using the Keras model provided.
I0713 10:53:17.262778 139797621028608 tf_logging.py:115] Using config: {'_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_task_type': 'worker', '_global_id_in_cluster': 0, '_is_chief': True, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f214e0a7210>, '_evaluation_master': '', '_save_checkpoints_steps': None, '_keep_checkpoint_every_n_hours': 10000, '_service': None, '_num_ps_replicas': 0, '_tf_random_seed': None, '_master': '', '_device_fn': None, '_num_worker_replicas': 1, '_task_id': 0, '_log_step_count_steps': 100, '_model_dir': 'model_20m', '_train_distribute': <tensorflow.contrib.distribute.python.one_device_strategy.OneDeviceStrategy object at 0x7f214e0a71d0>, '_save_summary_steps': 100}
2018-07-13 10:53:17.263159: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1471] Adding visible gpu devices: 0
2018-07-13 10:53:17.263193: I tensorflow/core/common_runtime/gpu/gpu_device.cc:952] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-07-13 10:53:17.263199: I tensorflow/core/common_runtime/gpu/gpu_device.cc:958]      0
2018-07-13 10:53:17.263218: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   N
2018-07-13 10:53:17.263362: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1084] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 11162 MB memory) -> physical GPU (device: 0, name: TITAN Xp, pci bus id: 0000:01:00.0, compute capability: 6.1)
2018-07-13 10:53:17.354058: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1471] Adding visible gpu devices: 0
2018-07-13 10:53:17.354097: I tensorflow/core/common_runtime/gpu/gpu_device.cc:952] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-07-13 10:53:17.354103: I tensorflow/core/common_runtime/gpu/gpu_device.cc:958]      0
2018-07-13 10:53:17.354108: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   N
2018-07-13 10:53:17.354218: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1084] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 11162 MB memory) -> physical GPU (device: 0, name: TITAN Xp, pci bus id: 0000:01:00.0, compute capability: 6.1)
2018-07-13 10:53:19.087784: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1471] Adding visible gpu devices: 0
2018-07-13 10:53:19.087843: I tensorflow/core/common_runtime/gpu/gpu_device.cc:952] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-07-13 10:53:19.087848: I tensorflow/core/common_runtime/gpu/gpu_device.cc:958]      0
2018-07-13 10:53:19.087852: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   N
2018-07-13 10:53:19.088024: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1084] Created TensorFlow device (/device:GPU:0 with 11162 MB memory) -> physical GPU (device: 0, name: TITAN Xp, pci bus id: 0000:01:00.0, compute capability: 6.1)
I0713 10:53:19.088659 139797621028608 tf_logging.py:115] Benchmark run: {'machine_config': {'cpu_info': {'cpu_info': 'Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz', 'mhz_per_cpu': 3700.0, 'num_cores': 12}, 'gpu_info': {'count': 1, 'model': u'TITAN Xp'}, 'memory_total': 67480199168, 'memory_available': 60690792448}, 'run_date': '2018-07-13T17:53:17.670323Z', 'tensorflow_version': {'git_hash': 'v1.9.0-rc0-2716-ge1436b2952', 'version': '1.9.0-rc0'}, 'dataset': {'name': 'ml-20m'}, 'tensorflow_environment_variables': [], 'run_parameters': [{'long_value': 256, 'name': 'batch_size'}, {'name': 'hr_threshold', 'string_value': 'None'}, {'long_value': 8, 'name': 'number_factors'}, {'long_value': 2, 'name': 'train_epochs'}], 'model_name': 'recommendation'}
I0713 10:53:19.088999 139797621028608 tf_logging.py:115] Starting a training cycle: 1/2
I0713 10:55:52.807188 139797621028608 tf_logging.py:115] Calling model_fn.
I0713 10:55:53.052728 139797621028608 tf_logging.py:115] Done calling model_fn.
I0713 10:55:53.089109 139797621028608 tf_logging.py:115] Create CheckpointSaverHook.
[libprotobuf FATAL external/protobuf_archive/src/google/protobuf/message_lite.cc:68] CHECK failed: (byte_size_before_serialization) == (byte_size_after_serialization): tensorflow.GraphDef was modified concurrently during serialization.
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  CHECK failed: (byte_size_before_serialization) == (byte_size_after_serialization): tensorflow.GraphDef was modified concurrently during serialization.
Aborted (core dumped)

**Comment by christ1ne** [comment]
Is there any update on this issue? Thanks!

**Comment by robieta** [comment]
The current version in master uses TFRecords to serialize each example, which turns out to be very CPU expensive. I got around this by using multiprocessing, but the memory overhead was much higher than I realized. Overall it seems that some of the design choices turned out to be non-ideal.

I'm currently ripping out the entire input pipeline and rewriting it which should help significantly. I'll update as I move closer to submitting.
